Modules/_sqlite/connection.c

```diff
@@ -432,48 +437,81 @@ free_callback_contexts(pysqlite_Connection *self)
 static void
 remove_callbacks(sqlite3 *db)
 {
-    sqlite3_trace_v2(db, SQLITE_TRACE_STMT, 0, 0);
+    /* None of these APIs can fail, as long as they are given a valid
+     * database pointer. */
+    assert(db != NULL);
+    int rc = sqlite3_trace_v2(db, SQLITE_TRACE_STMT, 0, 0);
+    assert(rc == SQLITE_OK), (void)rc;
+
     sqlite3_progress_handler(db, 0, 0, (void *)0);
-    (void)sqlite3_set_authorizer(db, NULL, NULL);
+
+    rc = sqlite3_set_authorizer(db, NULL, NULL);
+    assert(rc == SQLITE_OK);
 }
 
-static void
+static int
 connection_close(pysqlite_Connection *self)
 {
+    int rc = 0;
     if (self->db) {
         if (self->autocommit == AUTOCOMMIT_DISABLED &&
             !sqlite3_get_autocommit(self->db))
         {
-            /* If close is implicitly called as a result of interpreter
-             * tear-down, we must not call back into Python. */
-            if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
-                remove_callbacks(self->db);
+            if (connection_exec_stmt(self, "ROLLBACK") < 0) {
+                rc = -1;
             }
-            (void)connection_exec_stmt(self, "ROLLBACK");
         }
 
-        free_callback_contexts(self);
-
         sqlite3 *db = self->db;
         self->db = NULL;
 
         Py_BEGIN_ALLOW_THREADS
-        int rc = sqlite3_close_v2(db);
-        assert(rc == SQLITE_OK), (void)rc;
+        /* The v2 close call always returns SQLITE_OK if given a valid database
+         * pointer (which we do), so we can safely ignore the return value */
+        (void)sqlite3_close_v2(db);
         Py_END_ALLOW_THREADS
+
+        free_callback_contexts(self);
     }
+    return rc;
 }
 
 static void
-connection_dealloc(pysqlite_Connection *self)
+connection_finalize(PyObject *self)
 {
-    PyTypeObject *tp = Py_TYPE(self);
-    PyObject_GC_UnTrack(self);
-    tp->tp_clear((PyObject *)self);
+    pysqlite_Connection *con = (pysqlite_Connection *)self;
+    PyObject *exc = PyErr_GetRaisedException();
+
+    /* If close is implicitly called as a result of interpreter
+     * tear-down, we must not call back into Python. */
+    PyInterpreterState *interp = PyInterpreterState_Get();
+    int teardown = _Py_IsInterpreterFinalizing(interp);
+    if (teardown && con->db) {
+        remove_callbacks(con->db);
```

Comment:
Yeah, I think that it's better to remove callbacks before calling close. It makes the code easier to understand.

Reply:
Yes, and it is important to also remove them before executing the ROLLBACK.

Reply:
I hope that users will start to always close their sqlite database connection so we can avoid this madness during Python finalization!

Reply:
Ditto!